src/fidesops/service/authentication/authentication_strategy_oauth2.py

```diff
@@ -0,0 +1,291 @@
+                token_request,
+                connection_config.secrets,
+            )
+            # ignore errors so we can return the error message in the response
+            response = client.send(prepared_request, ignore_errors=True)
+            json_response = response.json()
+        except Exception as exc:
+            logger.error(
+                "Error occurred during the %s request for %s: %s",
+                action,
+                NotPii(connection_config.key),
+                str(exc),
+            )
+            raise OAuth2TokenException(
+                f"Error occurred during the {action} request for {connection_config.key}: {str(exc)}"
+            )
+
+        return json_response
+
+    def _validate_and_store_response(
+        self,
+        db: Optional[Session],
+        response: Dict[str, Any],
+        connection_config: ConnectionConfig,
+    ) -> str:
+        """
+        Persists and returns the new access token.
+        Also updates the refresh token if one is provided.
+        """
+
+        access_token = response.get("access_token")
+
+        # error, error_description, and error_uri are part of the OAuth2 spec
+        if access_token is None:
+            error_message = " ".join(
+                filter(
+                    None,
+                    (
+                        f"Unable to retrieve token for {connection_config.key} ({response.get('error')}).",
+                        response.get("error_description"),
+                        response.get("error_uri"),
+                    ),
+                )
+            )
+            logger.error(error_message)
+            raise OAuth2TokenException(error_message)
+
+        data = {"access_token": access_token}
+
+        # The authorization server MAY issue a new refresh token, in which case
+        # the client MUST discard the old refresh token and replace it with the
+        # new refresh token.
+        #
+        # https://datatracker.ietf.org/doc/html/rfc6749#section-6
+
+        refresh_token = response.get("refresh_token")
+        if refresh_token:
+            data["refresh_token"] = refresh_token
+
+        # If omitted, the authorization server SHOULD provide the
+        # expiration time via other means or document the default value.
+        #
+        # https://datatracker.ietf.org/doc/html/rfc6749#section-5.1
+        #
+        # This alternate way of specifying the expiration is handled
+        # by the optional expires_in field of the OAuth2AuthenticationConfiguration
+
+        expires_in = response.get("expires_in") or self.expires_in
+        if expires_in:
+            data["expires_at"] = int(datetime.utcnow().timestamp()) + expires_in
+
+        # persist new tokens to the database
+        # ideally we use a passed in database session but we can
+        # get the session from the connection_config as a fallback
+        if not db:
+            db = Session.object_session(connection_config)
```

Comment:
The `Session.merge` approach I was using before didn't work. I would call an `update` on `merged_connection_config` using the new db session but `connection_config` on the original session did not receive the update. The result was that the updated secrets wouldn't persist and the token refresh would be attempted for each request. This new approach uses the existing session so changes made are available on subsequent `AuthenticatedClient` requests.